### ISSUE TITLE
Support for multiple storage directories

### ISSUE THREAD

**Issue by lfrancke** [stackabletech/hdfs-operator#274]
Currently, only a single storage directory is supported. We need to support an arbitrary number.
While we're also want to support heterogenous storage tagging.
CRD design
```
  dataNodes:
    roleGroups:
      default:
        replicas: 1
        config:
          resources:
            storage:
              data: # We need to overwrite the data pvcs coming from the default value
                count: 0
              my-disks:
                capacity: 100Gi
                count: 5
                hdfsStorageType: Disk
              my-ssds:
                capacity: 10Gi
                storageClass: premium
                count: 3
                hdfsStorageType: Ssd
```

By defaulting to 
```
              data
                capacity: 5Gi
                count: 1
                hdfsStorageType: Disk
```
this is a non-breaking change for the CRD. Because we name the pvcs `data`, `data-1`, `data-2` etc. even the pvc names do not change, so the whole feature should be non-breaking

- [ ] Multiple pvcs supported
- [ ] `hdfsStorageType` is passed to hdfs
- [ ] CRD change is non-breaking
- [ ] pvc name change is non-breaking
- [ ] Tests added
- [ ] Documentation added

**Comment by sbernauer** [comment]
@soenkeliebau @lfrancke I assume we only need this for the datanodes?

**Comment by soenkeliebau** [comment]
I'd say so, yes.
iirc `dfs.datanode.data.dir` is the setting for this, so would only target datanodes..

**Comment by lfrancke** [comment]
There is an equivalent setting for the namenode. We might want to do that at the same time while we're on it.

**Comment by sbernauer** [comment]
Do the pvcs need to be identical? Currently i was only thinking about a `numberOfDataPvcs: u16` counter and that's it.
But reading a bit i stumbled across
> Namenode : 2 directories are enough for backing up namenode metadata in case of any crash of namenode. Usually 1st disk should be local disk and prefer 2nd disk as network storage[san/nas](Just incase local machine goes down you can have backup of namenode metadata on network storage). If you do not have network storage then 2 local disk are fine.

Which let me to the question if we want to support different storageclasses (basically a complete PvcConfig) for every single pvc instance? Currently the user can specify how the `data` pvc should look like and that get's replicated `numberOfDataPvcs` times
Use-case would be to use two different storage providers for redundany. But i think we can also solve (maybe even better) this use-case with multiple namenode rolegroups that use different storageclasses. Also i think that hdfs will be pretty f*cked when one of the two storage providers crashes.

TLDR: Going with the one pvs definition replicated `numberOfDataPvcs` times approach

**Comment by sbernauer** [comment]
Reading about `dfs.namenode.name.dir`, it seems to me that it is meant for redundancy and not to improve performance
> Determines where on the local filesystem the DFS name node should store the name table(fsimage). If this is a comma-delimited list of directories then the name table is replicated in all of the directories, for redundancy. 

Having this in mind is it really a good idea to increase the availability of the namenode this way? With the current approach all the pvcs will be provided from a single storage provider anyway - thus not really increasing the availability. I would think that the correct approach would be to increase the number of datanodes - which resource separation, health-checks and all the other good things. But I'm far from a hdfs expert.

**Comment by soenkeliebau** [comment]
We might need to look into [storage types and maybe storage policies](https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/ArchivalStorage.html) to answer this fully I'm afraid.

Haven't looked at this in ages tbh, but iirc we'd need to be able to specify per volume the storage type that it has.

Then again, that could be accomodated by just having two blocks with 

```yaml
storage:
  - storageClass: default
    numberOfPvs: 5
    storageType: disk
  - storageClass: premium
    numberOfPvs: 2
    storageType: ssd
```

and maybe make the storageType optional and default it to `disk` so we'd get 

```yaml
storage:
  - storageClass: default
    numberOfPvs: 5
  - storageClass: premium
    numberOfPvs: 2
    storageType: ssd
```

**Comment by sbernauer** [comment]
Good to know.
Maybe we should implement multtiple pvcs in a Version 1 and maybe later on add the storage policies if needed. But we should consider the storage policies in the CRD already, so we can add them non-breaking later on

**Comment by soenkeliebau** [comment]
Agreed

**Comment by sbernauer** [comment]
And i assume that a single node should support multiple storage types at once? (If not, things should be okay the way they are now)

**Comment by sbernauer** [comment]
Normally I would suggest the following:
```
  dataNodes:
    roleGroups:
      default:
        replicas: 1
        config:
          resources:
            storage:
              - capacity: 100Gi
                hdfsStorageType: disk # field ignored for now
                numberOfPvcs: 5
              - capacity: 10Gi
                hdfsStorageType: ssd # # field ignored for now
                storageClass: premium
                numberOfPvcs: 2
```
Problem here is, that we can't use `Vec` with the Fragement/Valdiation/Config merging machinery AFAIK.

What works is
```
           resources:
            storage:
              disk: # pvcs will still be named data to not force users to rename pvcs.
                numberOfPvcs: 5
                pvc:
                  capacity: 100Gi
              ssd: # supported later on
                numberOfPvcs: 3
                pvc:
                  capacity: 10Gi
                  storageClass: premium
```

What do you think about the latter solution?

**Comment by sbernauer** [comment]
Disregard my last comment, the current proposal is in the Ticket description

**Comment by soenkeliebau** [comment]
Looks good to me in principle, just a few comments:
- I think Hadoop uses all caps for the classes, so DISK and SSD - we should honor that
- The ticket doesn't explicitly state this (or I missed it), but I think we should allow ommitting the `hdfsStorageType` and default to DISK in that case most users won't care about this I think
- One thing I noticed when testing the spike is, that when changing the disk setup on an existing cluster the operator fails to apply this, because it is not allowed to patch the pvc section of the statefulset (might change in time, but currently just not supported: https://github.com/kubernetes/kubernetes/issues/68737 https://github.com/kubernetes/enhancements/pull/3412 )
  - Either we find a simple solution to notice this and delete the statefulset in that case
  - Or we document that for this specific action the user has to manually delete the statefulset for the datanodes to get the operator "unstuck"
  - More long term, we should add code to the orphaned objects module in operator-rs to "force" these scenarios

**Comment by sbernauer** [comment]
Regarding
1. We write the type already uppercase into the config (https://github.com/stackabletech/hdfs-operator/blob/a7a4cbefa64a84982b23d6dcc76a5f7b5c3dd504/rust/crd/src/lib.rs#L1351). Or do you mean exposing it uppercase in the crd (I would not recommend that)?
2. We should already default to disk: https://github.com/stackabletech/hdfs-operator/blob/a7a4cbefa64a84982b23d6dcc76a5f7b5c3dd504/rust/crd/src/lib.rs#L641-L646 and https://github.com/stackabletech/hdfs-operator/blob/a7a4cbefa64a84982b23d6dcc76a5f7b5c3dd504/rust/crd/src/lib.rs#L626
3. I haven't tried out any disk migration during the Spike. That's a thing we should research during this task, thx

**Comment by soenkeliebau** [comment]
No further questions your honor :)

**Comment by sbernauer** [comment]
Perfect Sir :+1:

**Comment by sbernauer** [comment]
Blocked by https://github.com/stackabletech/hdfs-operator/issues/288
